### ISSUE TITLE
Pattern matching doesn't check up the specialization chain

### ISSUE THREAD

**Issue by antroseco** [DaveDuck321/GrapheneLang#147]
```rust
typedef [T] Foo : {}
typedef [T] Bar : Foo<T>

function [T] takes_bar : (_ : Bar<T>) -> void = {}

function main : () -> int = {
    let foo : Foo<int> = {};

    // This doesn't compile, but it would if Foo and Bar weren't templates. 
    takes_bar(foo);

    return 0;
}
```

Seems inconsistent to me, but the types overview doc isn't 100% clear about what should happen here. If `Foo` and `Bar` aren't templates then the doc states that it should work (and it does), because

> Name1, Name2, and TypeThatWeWantToAlias are equivalent and can be used interchangeably in any context.

which kinda implies to me that they are all equivalent with each other.

But for specialized types

> When specialized with a type equivalent to SpecializationType, Generic\<SpecializationType\> is equivalent to OtherType.

But it's not clear to me whether that implies that _OtherType is equivalent to Generic\<SpecializationType\>_.

So does _A equivalent to B_ also imply _B equivalent to A_?

https://github.com/DaveDuck321/GrapheneLang/blob/main/docs/types_overview.md#aliased-type

**Comment by DaveDuck321** [comment]
Yeah, these definitions conflict. At a minimum there is a bug in the documentation 🤒. 
>But it's not clear to me whether that implies that OtherType is equivalent to Generic<SpecializationType>.
>
> So does A equivalent to B also imply B equivalent to A?

It says they can be "used interchangeably in any context" so I think this is implied. They *should* be equivalent to one-another.


The problem in the example is pattern matching fails. I'm sure if you called `takes_bar<Foo<int>>(foo)` it'd work.

**Comment by DaveDuck321** [comment]
So for a resolution. We can:
1) Change the documention: "used interchangeably in any strongly typed context"
2) Change the implementation: somehow search up the alias chain.

The implementation change introduces some challenges tho:
1) what to do if there's two functions, one takes foo one takes bar... Im assuming this has to be an error now.
2) since we can no longer use the alias information from the concrete type, we have to exhaustivly search the alias chain for every way foo could become bar (which would require speculative sfinae 🤮

I vote we update the documention and require the manual specialization in this case?

**Comment by antroseco** [comment]
I've got an idea - what if we _always_ follow the aliases down the chain, for both types, and then do the pattern matching? 

If that doesn't work then specifying the specialization manually isn't too bad, although some code won't be as nice (I was looking at the `slice_to` functions: we have one for each of `View` and `Span`, but if we fix this we could just write it once for `_SpanImpl`)

**Comment by DaveDuck321** [comment]
> I've got an idea - what if we always follow the aliases down the chain, for both types, and then do the pattern matching?

The problem with this is we don't know both types. In your example, we know the type of foo because we've instantiated it but we can't instantiate bar until after the pattern match.


The first problem: our types can be many-to-one.
```c
typedef[@V] Foo : {}
typedef Bar<0> : Foo<false>;
typedef Bar<1> : Foo<true>;
typedef Bar<2> : Foo<false>;

function [V] takes_foo : (_ : Foo<@V>) -> void = {}
function [V] takes_bar : (_ : Bar<@V>) -> void = {}

function main : () -> int = {
    // *Should* work since we can see that Bar<0> is equivalent to Foo<false>.
    takes_foo(make<Bar<0>>());

   // Doesn't/ shouldn't work... What do we match as `T`? `Bar<0>` or `Bar<2>`
   takes_bar(make<Foo<false>>())
}
```

The second problem: what if its really expensive to find valid types.
```c
typedef[@V] Foo : {}
typedef[@V] Bar : typeif<Foo<true>, IsInt64Max<@V>>;

function [V] takes_bar : (_ : Bar<@V>) -> void = {}

function main : () -> int = {
   // Here Foo is equivalent to Bar<0xffffffffffffffff> and we want to pattern match Bar's argument
   // But... We've never instantiated Bar with this argument before. Would the compiler just have
  // to search for an argument to Bar that would make the SFINAE work?
   takes_bar(make<Foo>())
}
```
